Starting from version 1.0.0, a union of Generators is (sometimes) treated as if it has no ReturnType

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

Starting from version 1.0.0, `yield from` a union on `Generator` types generates the error `Function does not return a value  [func-returns-value]`, and attempting to use the return reveals that Mypy is, in fact, treating it as if it's None.

There are ways around this; it appears that the inference is a little smarter at function boundaries than in-line, so I've just defined a bunch of helper functions to work around it for now.

**To Reproduce**

```python
from typing import Generator, TypeVar

T = TypeVar("T")


def my_iter(arg: Generator[int, None, T] | Generator[str, None, T]) -> Generator[int | str, None, T]:
    return (yield from arg)
```
(https://mypy-play.net/?mypy=0.991&python=3.11&gist=5c3183a0d4d1c02161211b01654d519f)

(The repro is working with `YieldType`, but in my code it's changing `ReturnType`. It breaks either way around.)

**Actual Behavior**

On post-1.0.0 versions:
```
main.py:7: error: Function does not return a value  [func-returns-value]
main.py:7: error: Incompatible return value type (got "None", expected "T")  [return-value]
Found 2 errors in 1 file (checked 1 source file)
```
Pre-1.0.0, produces no output.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -992,8 +992,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # AwaitableGenerator, Generator: tr is args[2].
             return return_type.args[2]
         else:
-            # Supertype of Generator (Iterator, Iterable, object): tr is any.
-            return AnyType(TypeOfAny.special_form)
+            # We have a supertype of Generator (Iterator, Iterable, object)
+            # Treat `Iterator[X]` as a shorthand for `Generator[X, Any, None]`.
+            return NoneType()
 
     def visit_func_def(self, defn: FuncDef) -> None:
         if not self.recurse_into_functions:

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5943,17 +5943,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
         # Determine the type of the entire yield from expression.
         iter_type = get_proper_type(iter_type)
-        if isinstance(iter_type, Instance) and iter_type.type.fullname == "typing.Generator":
-            expr_type = self.chk.get_generator_return_type(iter_type, False)
-        else:
-            # Non-Generators don't return anything from `yield from` expressions.
-            # However special-case Any (which might be produced by an error).
-            actual_item_type = get_proper_type(actual_item_type)
-            if isinstance(actual_item_type, AnyType):
-                expr_type = AnyType(TypeOfAny.from_another_any, source_any=actual_item_type)
-            else:
-                # Treat `Iterator[X]` as a shorthand for `Generator[X, None, Any]`.
-                expr_type = NoneType()
+        expr_type = self.chk.get_generator_return_type(iter_type, is_coroutine=False)
 
         if not allow_none_return and isinstance(get_proper_type(expr_type), NoneType):
             self.chk.msg.does_not_return_value(None, e)

--- a/mypyc/test-data/run-generators.test
+++ b/mypyc/test-data/run-generators.test
@@ -246,12 +246,12 @@ assert run_generator(another_triple()()) == ((1,), None)
 assert run_generator(outer()) == ((0, 1, 2, 3, 4), None)
 
 [case testYieldThrow]
-from typing import Generator, Iterable, Any
+from typing import Generator, Iterable, Any, Union
 from traceback import print_tb
 from contextlib import contextmanager
 import wrapsys
 
-def generator() -> Generator[int, None, int | None]:
+def generator() -> Generator[int, None, Union[int, None]]:
     try:
         yield 1
         yield 2

--- a/mypyc/test-data/run-generators.test
+++ b/mypyc/test-data/run-generators.test
@@ -251,7 +251,7 @@ from traceback import print_tb
 from contextlib import contextmanager
 import wrapsys
 
-def generator() -> Iterable[int]:
+def generator() -> Generator[int, None, int | None]:
     try:
         yield 1
         yield 2
@@ -263,6 +263,8 @@ def generator() -> Iterable[int]:
             print('caught exception with value ' + s)
         else:
             print('caught exception without value')
+        return 0
+    return None
 
 def no_except() -> Iterable[int]:
     yield 1
@@ -354,11 +356,11 @@ with ctx_manager() as c:
     raise Exception
   File "native.py", line 10, in generator
     yield 3
-  File "native.py", line 30, in wrapper
+  File "native.py", line 31, in wrapper
     return (yield from x)
   File "native.py", line 9, in generator
     yield 2
-  File "native.py", line 30, in wrapper
+  File "native.py", line 31, in wrapper
     return (yield from x)
 caught exception without value
 caught exception with value some string

--- a/mypyc/test-data/run-generators.test
+++ b/mypyc/test-data/run-generators.test
@@ -263,7 +263,6 @@ def generator() -> Iterable[int]:
             print('caught exception with value ' + s)
         else:
             print('caught exception without value')
-        return 0
 
 def no_except() -> Iterable[int]:
     yield 1

--- a/test-data/unit/check-statements.test
+++ b/test-data/unit/check-statements.test
@@ -2253,6 +2253,16 @@ from typing import Generator, Union, Iterable, Optional
 
 def foo(arg: Union[Generator[int, None, bytes], Iterable[str]]) -> Generator[Union[int, str], None, Optional[bytes]]:
     return (yield from arg)
+
+def bar(arg: Generator[str, None, str]) -> Generator[str, None, str]:
+    return foo(arg)  # E: Incompatible return value type (got "Generator[Union[int, str], None, Optional[bytes]]", expected "Generator[str, None, str]")
+
+def launder(arg: Iterable[str]) -> Generator[Union[int, str], None, Optional[bytes]]:
+    return foo(arg)
+
+def baz(arg: Generator[str, None, str]) -> Generator[Union[int, str], None, Optional[bytes]]:
+    # this is unsound, the Generator return type will actually be str
+    return launder(arg)
 [builtins fixtures/tuple.pyi]
 
 [case testNoCrashOnStarRightHandSide]

--- a/test-data/unit/check-statements.test
+++ b/test-data/unit/check-statements.test
@@ -2242,18 +2242,16 @@ def foo(arg: Union[Generator[int, None, T], Generator[str, None, T]]) -> Generat
 [case testYieldFromInvalidUnionReturn]
 from typing import Generator, Union
 
-class T: pass
-class R: pass
+class A: pass
+class B: pass
 
-def foo(arg: Union[T, R]) -> Generator[Union[int, str], None, T]:
-    return (yield from arg) # E: "yield from" can't be applied to "Union[T, R]"
+def foo(arg: Union[A, B]) -> Generator[Union[int, str], None, A]:
+    return (yield from arg) # E: "yield from" can't be applied to "Union[A, B]"
 
 [case testYieldFromUnionOfGeneratorWithIterableStr]
 from typing import Generator, Union, Iterable, Optional
 
-class T: pass
-
-def foo(arg: Union[Generator[int, None, T], Iterable[str]]) -> Generator[Union[int, str], None, Optional[T]]:
+def foo(arg: Union[Generator[int, None, bytes], Iterable[str]]) -> Generator[Union[int, str], None, Optional[bytes]]:
     return (yield from arg)
 [builtins fixtures/tuple.pyi]
 

--- a/test-data/unit/check-statements.test
+++ b/test-data/unit/check-statements.test
@@ -2265,6 +2265,17 @@ def baz(arg: Generator[str, None, str]) -> Generator[Union[int, str], None, Opti
     return launder(arg)
 [builtins fixtures/tuple.pyi]
 
+[case testYieldIteratorReturn]
+from typing import Iterator
+
+def get_strings(foo: bool) -> Iterator[str]:
+    if foo:
+        return ["foo1", "foo2"]  # E: No return value expected
+    else:
+        yield "bar1"
+        yield "bar2"
+[builtins fixtures/tuple.pyi]
+
 [case testNoCrashOnStarRightHandSide]
 x = *(1, 2, 3)  # E: can't use starred expression here
 [builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-statements.test
+++ b/test-data/unit/check-statements.test
@@ -85,7 +85,7 @@ def f() -> Generator[int, None, None]:
 from typing import Iterator
 def f() -> Iterator[int]:
     yield 1
-    return "foo"
+    return "foo" # E: No return value expected
 [out]
 
 
@@ -2230,6 +2230,32 @@ class B: pass
 
 def foo(x: int) -> Union[Generator[A, None, None], Generator[B, None, None]]:
     yield x  # E: Incompatible types in "yield" (actual type "int", expected type "Union[A, B]")
+
+[case testYieldFromUnionOfGenerators]
+from typing import Generator, Union
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Generator[str, None, T]]) -> Generator[Union[int, str], None, T]:
+    return (yield from arg)
+
+[case testYieldFromInvalidUnionReturn]
+from typing import Generator, Union
+
+class T: pass
+class R: pass
+
+def foo(arg: Union[T, R]) -> Generator[Union[int, str], None, T]:
+    return (yield from arg) # E: "yield from" can't be applied to "Union[T, R]"
+
+[case testYieldFromUnionOfGeneratorWithIterableStr]
+from typing import Generator, Union, Iterable, Optional
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Iterable[str]]) -> Generator[Union[int, str], None, Optional[T]]:
+    return (yield from arg)
+[builtins fixtures/tuple.pyi]
 
 [case testNoCrashOnStarRightHandSide]
 x = *(1, 2, 3)  # E: can't use starred expression here

